coderd/database/migrations/000278_workspace_next_start_at.up.sql

```diff
@@ -0,0 +1,59 @@
+ALTER TABLE ONLY workspaces ADD COLUMN IF NOT EXISTS next_start_at TIMESTAMPTZ DEFAULT NULL;
+
+CREATE INDEX workspace_next_start_at_idx ON workspaces USING btree (next_start_at) WHERE (deleted=false);
+
+CREATE FUNCTION nullify_workspace_next_start_at() RETURNS trigger
+	LANGUAGE plpgsql
+AS $$
+DECLARE
+BEGIN
+	IF (NEW.autostart_schedule <> OLD.autostart_schedule) AND (NEW.next_start_at = OLD.next_start_at) THEN
+		UPDATE workspaces
+		SET next_start_at = NULL
+		WHERE id = NEW.id;
+	END IF;
+	RETURN NEW;
+END;
+$$;
+
+CREATE TRIGGER trigger_update_workspaces_schedule
+	AFTER UPDATE ON workspaces
+	FOR EACH ROW
+EXECUTE PROCEDURE nullify_workspace_next_start_at();
```

Comment:
I'm thinking we also need a trigger for changes to a template time-til-dormant? Or do I misunderstand that interaction?

My thought process: if the time-til-dormant is reduced, we might've previously calculated a next start for the workspace which might no longer be valid because dormancy was sped up.

It's probably a fringe edge-case, but might be hard to debug customer reports if they run into it.

I'm thinking this applies to template allowed days changing as well. Tomorrow might've been blocked but is now unblocked yet we computed the autostart for the day after.

I'm wondering if there are other such interactions between template and workspace we should cover as well?

Reply:
I think I agree with you here, I'll try and think if there are any other similar interactions 👍

Reply:
> I'm thinking we also need a trigger for changes to a template time-til-dormant? Or do I misunderstand that interaction?

> My thought process: if the time-til-dormant is reduced, we might've previously calculated a next start for the workspace which might no longer be valid because dormancy was sped up.


After digging into this a little bit more, I don't think this is an issue.

I had forgotten to include `dormant_at IS NOT NULL` in `autostart` part of the `GetWorkspacesEligibleForTransition` SQL query, so I've added that now (it already existed in the coderd lifecycle executor side).

Instead of handling changes to `time_til_dormancy`, I've instead modified the existing trigger to check if the workspace is dormant. When a workspace is marked as dormant, the workspace's `next_start_at` is made NULL.

> I'm thinking this applies to template allowed days changing as well. Tomorrow might've been blocked but is now unblocked yet we computed the autostart for the day after.

This edge case has been handled in the `EnterpriseTemplateScheduleStore`